### DISCUSSION TITLE
UI Overhaul in version 3.x.x

### DISCUSSION THREAD

**Discussion by mebjas** [mebjas/html5-qrcode#213] [Ideas]
Hi everyone, I'm going to start working on a UI overhaul for this library - particularly the end to end scanner which can be directly integrated to any existing website for easy code scanning.

So far the UI was built based on my intuition, but I am not a qualified UXR/UXD and may have a lot of bias. In the next version I'd really appreciate some UXR/UXD inputs on the design, so we can address discoverability, comprehension, A11y & internationalization issues that exist in the library currently.

I'm super happy to discuss and implement ideas so it's great experience for both developers who use this library and the end users scanning the codes out in the world.

Some of the very early ideas I'd like to fix or improve are tracked in https://github.com/mebjas/html5-qrcode/issues/207

**Comment by ROBERT-MCDOWELL** [comment]
maybe start with a code condition checking if an extra css file exists and if not use the current webpage css?

**Comment by ygoe** [comment]
Oh cool, yes, improvement is definitely needed here. I discovered this library long ago, when it was still unreliable (reading garbage every other time). Then I noticed it switched to the new JS code backend and now it works much better. Today I went back to this topic and tried to build a simple app prototype around it. And [failed so far](https://github.com/mebjas/html5-qrcode/issues/219). The problem is that the camera view is completely uncontrollable. It's impossible to fit it into any given size constraints defined by an app UI. I spent the last couple hours hacking things away and redoing all the layout internally done by the library. But new problems appeared. Nothing lines up, and the layout and appearance is totally unreliable.

I was tempted to just throw all of the overlays out of the window and doing it myself. Trouble is, there's no way to tell when a code was discovered, that's the green bars around the scan box. So I still need those.

My top priority here is that the scanner view must be usable in a box of a fixed size. It must obey that size and fit itself into that, it must never change the size of that box or draw anything outside of it. Ideally it should cover the box and center itself in there. And when the window resizes, it should update its layout accordingly on its own.

It would be helpful if I could specify the QR scan box size (inside the shaded area) as relative to the minimum view length instead of absolute pixels. That lets it adapt to different screen sizes more easily. Again, when the window resizes, that number needs to be updated automatically. (This cannot be done now without stopping and restarting the scan video because that box size is a start parameter.)

I don't need any localisation because there's no text in that view anyway.

More events might be useful to be able to use custom detection indicators.

So I'm suspending my work on this now and see what's happening here. My prototype isn't time critical so I can wait and see what improvements come here.

On the API side: I'd like to get more information about the scanned QR code (raw bytes, error correction level, content mode, mask pattern). And I'd like to change the scan box size (within shaded area) on the fly (handy if you notice there are too many codes next to each other). Is it possible to reduce the set of supported codes to reduce the file size? That thing is huge and I just need QR codes, not a hundred others I'll never see (and that are irrelevant in my application). If my app says "scan that QR code" and then it finds a barcode, that's stupid.

**Reply by Isaias-Jose** [reply]
The layot ​​issue is a fact, I also came across this problem

**Comment by ROBERT-MCDOWELL** [comment]
yeah I noticed that the mobile orientation ruins the canvas too.

**Comment by mebjas** [comment]
Some of my pure UI thoughts https://www.figma.com/file/SdXeVf1pBoE9lcfjhYtFcd/Html5Qrcode?node-id=0%3A1

**Reply by damal1997** [reply]
Hello, please could you help, I am using Intascan library, but it is not working on the mobile browser.
Can I do it with your mebjas/html5-qrcode?
if yes, can you do a few tutorials, or give me a link or other which will allow me to realize my project.

**Reply by mebjas** [reply]
https://blog.minhazav.dev/QR-and-barcode-scanner-using-html-and-javascript/

Give this one a try! Hope this helps

**Reply by holmesatmewt** [reply]
@mebjas with the present version, how can we create scanners like in figma design file

**Comment by Dvdben** [comment]
It is impossible to download the library from an android phone

**Reply by mebjas** [reply]
what do you mean download the library from an android phone?

**Comment by nikowuesti** [comment]
Hey, I did not work with the library yet, but I will in the next months, because my rental-system needs a feature to scan items and it needs to be web based. So I will be invested in making the UI work with my design and here are my early ideas:

- Your new layout from Aug 1st is promising and the buttons we need are there
- Icons could be SVGs that are defined within the code (small and no extra resource and customizable color)
- As a developer I need to be able to customize the icons design and position (can be a CSS override)
- As a developer I need to be able to call these buttons with methods from the library
- As mentioned in a comment above, the wrapper size can be fixed and the scanner + UI inside should be responsive

**Reply by mebjas** [reply]
Sounds great to me!

Happy to discuss offline / online on the goals and review the PRs.

**Comment by philp** [comment]
Hi! Just stumbled across this great looking project while scoping for a current project. I'm a full stack developer, but with a heavy leaning into front-end UI/UX. I've been looking for an OSS side-project to contribute to, and think I could have a lot to offer.

**Reply by mebjas** [reply]
Sounds great to me!

Happy to discuss offline / online on the goals and review the PRs.

**Reply by philp** [reply]
Great. I'm just coming out of a busy patch, so going to take a bit of time to get familiar with the codebase and will then think about the best way to share my thoughts/ideas.

**Comment by Dvdben** [comment]
I mean, you need to scan the qrcode through your cell phone webcam.
Connecting the phone with the website. And I tried the library but nothing
works

**Reply by mebjas** [reply]
Does the demo page work - qrcode.minhazav.dev

**Comment by rodneystover** [comment]
Not sure if I'm posting this in the correct place. I am using the Html5Qrcode source successfully and was wondering how I would insert code to add a toggle to access on/off for the flashlight in my code.  I found the support for flashlight option here: [https://github.com/lafriks-fork/html5-qrcode/commit/23c8e76975b34947da2c783b366583c42c85edde](url)

This is my settings before I call the camera:

const config = { fps: 60,
				 qrbox: 275,
				 aspectRatio: 1.0,
			     experimentalFeatures: {
				     useBarCodeDetectorIfSupported: true
				 }
			   };	

html5QrCode.start({ facingMode: { exact: "environment"} }, config, qrCodeSuccessCallback);

**Reply by mebjas** [reply]
Planning to add support for this soon - please subscribe to https://github.com/mebjas/html5-qrcode/issues/129

**Comment by rodneystover** [comment]
I am not using the Html5QrcodeScanner  Ui I am using the  Html5Qrcode  Version
because I am coding a web app for my delivery drivers

**Comment by Dvdben** [comment]
How do you do it ?


Le jeudi 30 septembre 2021, rodneystover ***@***.***> a
écrit :

> I am not using the Html5QrcodeScanner Ui I am using the Html5Qrcode Version
> because I am coding a web app for my delivery drivers
>
>

**Comment by rodneystover** [comment]
Here is the javascript code I am using to make it work. I have it working
as a progressive web app on iphone 8.
I am able to scan use continuous scanning mode or a single barcode mode and
have a confirmation beep on successful scan. If the barcode has already
been scanned it does a different type of beep. I am just wanting to add a
flashlight toggle switch in case scanning in low light conditions.

<script>
//------------------- barcode scanner ----------------------//
/** define scanning mode **/
// c - continous
// s - single
var scanMode = '<?php echo $scanMode;?>';

/** define target div container **/
const html5QrCode = new Html5Qrcode("reader");

/** create delay to prevent repeat scanning of same barcode **/
var sleep = function (ms) {
let now = Date.now(), end = now + ms;
while (now < end) { now = Date.now(); }
};

/** Create beep sound **/
var audioCtx = new (window.AudioContext || window.webkitAudioContext ||
window.audioContext);
//All arguments are optional:
//duration of the tone in milliseconds. Default is 500
//frequency of the tone in hertz. default is 440
//volume of the tone. Default is 1, off is 0.
//type of tone. Possible values are sine, square, sawtooth, triangle, and
custom. Default is sine.
//callback to use on end of tone
function beep(duration, frequency, volume, type, callback) {
    var oscillator = audioCtx.createOscillator();
    var gainNode = audioCtx.createGain();
    oscillator.connect(gainNode);
    gainNode.connect(audioCtx.destination);
    if (volume){gainNode.gain.value = volume;}
    if (frequency){oscillator.frequency.value = frequency;}
    if (type){oscillator.type = type;}
    if (callback){oscillator.onended = callback;}
    oscillator.start(audioCtx.currentTime);
    oscillator.stop(audioCtx.currentTime + ((duration || 500) / 1000));
};

/** successful scan actions **/
const qrCodeSuccessCallback = (decodedText, decodedResult) => {
/** If barcode type is XXXXXXXXX retrieve only tracking number **/
const barcodeDetails = decodedText.split('\\n');
var howMany = barcodeDetails.length;
if (howMany != 1) {
decodedText = barcodeDetails[0];
}
if (scanMode === 's') { // single scan mode
/** Beep **/
beep(50, 2350, 1, 'square');
    /** Store scan result **/
$( "#scanResult" ).val(decodedText);
/** Stop scanning. **/
    html5QrCode.stop();
/** Clear the rendered surface. **/
html5QrCode.clear();
/** Return result to referring URL **/
$( "#submitscan" ).submit();
}
if (scanMode === 'c') { // continuous scan mode
    /** Check for duplicate scan **/
newScan = document.getElementById("scanResult").value;
var arr = newScan.split(",");
if (arr.indexOf(decodedText) !== -1) {  // duplicate scan
document.getElementById('scanMessage').innerHTML="Parcel Already Scanned";
beep(10, 150, .09, 'square');
sleep(100); // sound delay
beep(10, 150, .09, 'square');
sleep(2000); // 2 second delay
} else { // not duplicate scan
$( "#scanMessage" ).html('Parcel Scan Successful');
beep(50, 2350, 1, 'square');
/** Store scan result **/
if (!newScan) { // first scan
newScanValue = decodedText;
} else { // > than first scan so add comma delimiter
newScanValue = newScan+','+decodedText;
}
$( "#scanResult" ).val(newScanValue);
sleep(1000); // 1 second delay
}
}
};

/** define scanner default settings **/
const config = { fps: 60, // sets the framerate to 60 frame per second
qrbox: 275, // sets only 275 X 275 region of viewfinder to scannable, rest
shaded.
aspectRatio: 1.0,
     experimentalFeatures: {
     useBarCodeDetectorIfSupported: true
}
   };

/** load scanner using back camera **/
html5QrCode.start({ facingMode: { exact: "environment"} }, config,
qrCodeSuccessCallback);
/** User done scanning -> return results to referrer **/
function endScan() {
$( "#scanMessage" ).html('');
beep(20, 350, .09, 'sine');
/** End scanning. **/
    html5QrCode.stop();
/** Clear the rendered surface. **/
html5QrCode.clear();
/** Return results to referring URL **/
$( "#submitscan" ).submit();
}

/** User cancelled scan -> return to referrer **/
function cancelScan() {
$( "#scanMessage" ).html('');
beep(20, 350, .09, 'sine');
/** Stop scanning. */
    html5QrCode.stop();
/** Clear the rendered surface. */
html5QrCode.clear();
/** Redirect to referrer **/
location.href="<?php echo $referrer;?>?a=<?php echo $return_action;?>";
}

//------------------- end barcode scanner ----------------------//
</script>

**Comment by mebjas** [comment]
Everyone interested in this item - I strongly encourage you to subscribe to https://github.com/mebjas/html5-qrcode/issues/207 - I am thinking to take a stab at this for version 2.1.3 starting with svg icons.

Also, still looking for some UX help on the potential UX

**Comment by damal1997** [comment]
Thanks you very much.

God bless you.

**Comment by mng4** [comment]
I'm not core developer, but I would suggest if you find any way to remove the camera selection in case of multiple cameras. Only front and back would do. Else users might get confused if they have multiple cameras in back and not able to understand which one to select.

**Comment by AlfonsoML** [comment]
Besides internationalization, my main problem at the moment is that every style is inlined, a big step forward in order to allow us to customize the UI with little touches would be to use CSS classes, and even an event fired whenever the scanner changes state, so we could adjust other parts of the UI according to the state of the scanner.

In my case I would like to be able to overlay a top bar over the scanner, but at the moment it would also overlap the pre-scanner UI where the user must press the button to request permissions or switch camera. So I can't show properly my bar with the scanned info because it would overlap the pre-scanner text. I'm sure that I haven't explained my self properly, but the idea is: if the scanner used CSS classes and/or fired events, then I could do my job in one minute.

I'm going to try to implement this in a local fork, I guess that given that the plan is to rewrite all the UI then such PR would not be accepted for the project, otherwise I wouldn't mind to provide it.

**Reply by mebjas** [reply]
> I'm going to try to implement this in a local fork, I guess that given that the plan is to rewrite all the UI then such PR would not be accepted for the project, otherwise I wouldn't mind to provide it.

Thanks for bringing this up, +1, I think it would be helpful for both end users and devs.

Is there a general design patter for this kind of requirement?

**Reply by nikowuesti** [reply]
> Is there a general design patter for this kind of requirement?

Here are a few ideas in general:

- give all elements a meaningful class with the html5qr prefix, like "html5qr-container", "html5qr-container-overlay", "html5qr-scanner" and so on (these are just examples)
- do not use IDs at all, they will break if the plugin is initialized twice or if the developer has an ID with the same name in the code
- only use inline CSS when it comes to dynamic content, like transformations
- all other CSS should be put in the respective html5QrCode.css or so

**Reply by jasvvinder** [reply]
> > Is there a general design patter for this kind of requirement?
> 
> Here are a few ideas in general:
> 
> * give all elements a meaningful class with the html5qr prefix, like "html5qr-container", "html5qr-container-overlay", "html5qr-scanner" and so on (these are just examples)
> * do not use IDs at all, they will break if the plugin is initialized twice or if the developer has an ID with the same name in the code
> * only use inline CSS when it comes to dynamic content, like transformations
> * all other CSS should be put in the respective html5QrCode.css or so

Yes, These ideas are perfect that I would like to suggest too. For UI customization there should be separate CSS files based on class selectors.
As an front end developer I can contribute to this project.

**Comment by Mathias89** [comment]
Further suggestions:

- scanner should be headless and the ui should be optionally plugged in
- No IDs as already mentioned but include an selector like ".class-based", "[data-has-scanner]", "#id-if-only-used-once"
- No inline script and no inline css because CSP (content security policy) on many sites could be very strict (could also affect blob:)
- UI should adapt on decive orientation change
- Accessibility-wise allow only icons, icons with text, text only
- icons or buttons should be replaced by template-tag if configured
- all ui elements, like buttons, should be expandable by custom classes or at least a callback should be available before displaying ui to customise it without flickering
- naming classes by dashes is totally fine but BEM should also be considered
- each instance could have its own ui

**Comment by mebjas** [comment]
(WIP) Some major UI changes coming soon: https://scanapp.org/beta :)

**Reply by ygoe** [reply]
Interesting. Seems to work okay for a medium size QR code. When pointing at the Data Matrix code right next to it, the live image hangs for a moment, then it can decode Data Matrix codes multiple times. When pointing back to the QR code, nothing happens. As if there was nothing to see. Have to close and reopen the page to have it decode QR again. The first Data Matrix sight still breaks QR detection. (Tested in Brave/Android)

**Comment by mebjas** [comment]
Some of the proposals will be implemented in [scanapp-org/imagecodes](https://github.com/scanapp-org/imagecodes)

More info: https://github.com/mebjas/html5-qrcode/discussions/747
